Major update to token parsing and formatting

tsackton opened the pull request:
This is a major update to token parsing and string formatting. Introduces a new TokenParser class, that replaces StringFormatter. 

The main function is formatDisplayString, which takes a display string and formats the tokens appropriately, returning the formatted string. A token is a string with a format like:
`<(prefix)token:filterformat(suffix)>`

Prefix and suffix are optional, and can contain any character, including whitespace.
Token is required and must be letters only (upper and lower case are fine, but all tokens are converted to lowercase for processing, so endDate and enddate are not distinct). 
Filterformat is a filter and format string, which can contain any characters except `:()` or spaces. In can consist of a single string (a), or two or three strings separated by ;, e.g. (a;b) or (a;b;c). 
- A single string is interpreted as intermixed filter and format, but numeric slices must be first.
- In a pair of strings (a;b), 'a' is interpreted as intermixed filter and format, as before, while 'b' is interpreted as chainFormat, which is applied to all pieces of a whereabout chain (format applies only to the first piece)
- In a trio of strings (a;b;c), 'a' is interpreted as a filter, 'b' is interpreted as the format, and 'c' is interpreted as the chainFormat. 

This processing can be altered if needed / desired, but so far in testing seems to work as expected. 

formatDisplayString also takes a file object (should have a file.name and a file.frontmatter property), a targetDate (which is the date at which to format the string), and an overrides object, which is appended (and overwrites) the file metadata. The main uses so far of this are:
- if metadata.dateInfo exists, it is used instead of getting page dates from DateManager (can be used to format a string with different page date info, as in affiliation formatting)
- if metadata.name is a name object, it is passed to the Name Manager for formatting directly; if overrides contains a name property this will overwrite what is in metadata for formatting. 

Some bugs are not fully resolved. Current issues:
- casing only happens for the actual token, not the prefix; so e.g. Battle of Urlich Pass reports: "part of The [[Great War]]" which should probably be "Part of The [[Great War]]". The display string here is: "<(part of )partof:ty>" so maybe this is a display string bug not a code bug (e.g., should it be "<(Part of )partof:ty>"?
- The Mirror of Soul Trapping has a "RangeError: Maximum call stack size exceeded" which is probably an infinite loop somewhere. Vindristjarna has same error. 
- Trailing parentheses are being removed somewhere for page dated information and for affiliations, but not for secondary info line (apparently).

msackton committed:
typo in metadata

msackton committed:
Mostly working new whereabouts stuff

Commit:
add initial very incomplete attempt at a token parser class

msackton committed:
Initial work on improved formats for whereabout chains

msackton committed:
Merge branch 'new-whereabouts' of https://github.com/tsackton/taelgar into new-whereabouts

Commit:
handle range parsing in #parseTokenString

Commit:
token parser updates - still not working but general format is taking shape

Commit:
token parser updates; still needs functions but structure of parser is taking shape

Commit:
adding more to token parser

msackton committed:
some initial work on new name objects

msackton committed:
bugfixes

Commit:
theoretically nearly complete Token Parser class

Commit:
theoretically nearly complete Token Parser class

msackton committed:
a few more fixes

Commit:
token parser needs significant debugging but general structure is there

tsackton committed:
fix regexes to handle (prefix) and (suffix)

tsackton committed:
bugfixes to token parsing; now properly supports prefixes and suffixes

msackton committed:
Switch to token parser from string formatter

Commit:
first attempt at merge/reconciliation; not working

Commit:
working on merge/refactor; not debugged

Commit:
parseDisplayString renamed to formatDisplayString

Commit:
mostly debugged; some issues remain and further testing needed

Commit:
fix <name:null> format string bug

Commit:
more bugfixes

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+                        if (formatChars.includes(char)) format += char;
+                    }
+                } else if (filterFormatString.length === 2) {
+                    // we have a;b
```
This handling of a;b I think is slightly problematic. Specifically I don’t want 2;Q to be interpreted as “apply Q” to the entire chain but *not* to the current item. I think this cleverness in parsing is going to lead to unpredictable results where 2qr 2;qr and 2r;q all have confusingly different behaviors. 

It probably makes more sense to say that if you use a semi-colon, you can’t mix. That is:
2qrL is valid and treated as filter + format. 

But once you add a semi colon your choices are:
filter;format;chain
filter;format
format;chain

Some of this we might want to tweak as we play around. In particular, the chain filter and filter maybe should be reversed here, ie usually you want to apply the same to everything. 

The main use case is no link text in Originally from: so perhaps we should reverse which is chain and which is first. 

Either way I think the main point is that we should not allow filterformat;format

tsackton replied:
I think after fiddling a bit, the best option is to define:
filter - a filter string
format - a format string that applies everyone
firstFormat - a format string that overrides format for the first element in a chain

Then, we parse as:
- a single string is a filter+format string
- a string that has two parts (a;b) can be either a filter;format string or a format;firstFormat string
- a string that has three parts (a;b;c) is interpreted as a filter;format;firstFormat string

I think the logic is that normally you want the same format string across the whole chain, and it is really only a few special cases where you have a preposition or other link text in your display string that you want to override q with Q for just the first element.

tsackton replied:
We only apply the firstFormat to the first element in a chain if it exists, otherwise we just apply format.

tsackton replied:
Implemented in [adcf456](https://github.com/tsackton/taelgar/pull/25/commits/adcf456311d60ed891baf6a52a9e3a66519c89cd)

msackton replied:
Looks sensible. I had a feeling when playing around we'd want the defaults reversed.

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+
+        // parse the prefix, suffix, token, and format
+        let matches = input.matchAll(tokenRegex);
+        let tokenMatch = [...matches][0]
```
It shouldn’t be possible but it might be worth erroring out if there is more than one match.

tsackton replied:
[6d7c200](https://github.com/tsackton/taelgar/pull/25/commits/6d7c200be28263720a882ec69797d33d4c7c9bca) adds a check for match number.

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+        }
+
+        // all tokens are lower case //
+        token.token = token.token.toLowerCase()
```
This makes sense but there are a few mixed cases above (ie endStatus) - just for clarity might be worth cleaning up.

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+            let possibleAliasKey = token.token + "Alias"
+            let alias = this.#getParameterCaseInsensitive(metadata, possibleAliasKey)
+            let sourcePageType = NameManager.getPageType(metadata)
+            if (alias) {
```
You don’t need to check alias. It’s fine to pass the override with alias pointing to undefined.

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+
+            // start name options - can be a string or a name object //
+            case "name":
+                value = metadata.name ? metadata.name : file.name
```
This will break files that have an actual name attribute in metadata that doesn’t match the filename. (Ie some of the Cleenseau people who non-ascii names). We should only use the name instead of the filename is the name is a normalized name object)

msackton replied:
Note, it only breaks them if the name is not also in the set of aliases. So I guess a pretty edge case, in that normally you'd expect the name, if set to something other than the filename, to be an alias. Not obvious if this is worth fixing.

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+                value = metadata.subTypeOf
+                formatter = "name"
+                break
+            case "typeof":
```
This alias of main type to typeof means that a page with species and speciesAlias won’t work correctly for main type. I think you need to get the name object here so you can pass the right alias.

msackton replied:
Yeah, I tested this, it doesn't work.

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+                else if (metadata.gender == "male") value = "he/him"
+                else if (metadata.gender == "female") value = "she/her"
+                else if (metadata.gender) value = "they/them"
+                formatter = "none"
```
Might be worth making this use the casing format. I could imagine wanting pronouns:u

tsackton replied:
changed

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+                formatter = "none"
+                break;
+            case "age":
+                if (pageDateInfo.age == 0) value = "0 years"
```
Certainly not something that needs to be done in this pr but it would probably be better to have an age and population formatter rather than applying the format rules here.

tsackton replied:
Added age formatter in [80d4245](https://github.com/tsackton/taelgar/pull/25/commits/80d4245b81db5b9d48f71e1df1713bb044bd835e); not bothering with population formatter yet.

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+            formattedString += input.substring(lastIndex, tokenStartIndex);
+            if (this.debug) console.log("Token match: " + tokenMatch[0] + " at index " + tokenStartIndex + " which is " + (tokenStartIndex === 0));
+    
+            let token = this.#parseTokenString(tokenMatch[0], (tokenStartIndex === 0));
```
Is tokenStartIndex == 0 really going to be true when you have a string like `<ancestry> <species>` with no ancestry? Is first needs to be true when either (a) this is the first token or (b) all prior tokens had no values. 

You probably can just check the length of the result string.

tsackton replied:
fixed

msackton committed:
silly error on recursion in building the whereabout chain

msackton commented on _scripts/customJS/tokenParser.js:
```diff
@@ -0,0 +1,575 @@
+            let alias = this.#getParameterCaseInsensitive(metadata, possibleAliasKey)
+            let sourcePageType = NameManager.getPageType(metadata)
+            if (alias) {
+                name = NameManager.getNameObject(value, sourcePageType, undefined, { alias: alias })
```
This is also the wrong syntax for the call. There is no undefined here, you need just sourcePageType, {alias: alias}

Commit:
formatString has whitespace after failing to parse a token, trying to fix

Commit:
fix #parseTokenString isFirst bug

msackton committed:
fix aliases

Commit:
make pronouns use the casing formatter

msackton committed:
Use the override name if specified, otherwise always use
the file name, which is the "true target"

Commit:
fix ka bug; check for more than one token match in parser and skip processing if found

Commit:
forgot to append text after last token was parsed; fixed

tsackton commented:
The Mirror of Soul Trapping has a "RangeError: Maximum call stack size exceeded" which is probably an infinite loop somewhere. Vindristjarna has same error. --> fixed by [4698766](https://github.com/tsackton/taelgar/pull/25/commits/469876608577d4ad365615158ec4088a13ca561d)
Trailing parentheses are being removed somewhere for page dated information and for affiliations, but not for secondary info line (apparently). --> fixed by [eae071e](https://github.com/tsackton/taelgar/pull/25/commits/eae071e36d4a484cd7b462c7ca0d4bb4265644b8)

Commit:
fix typeOf missing formatter; add #getFormattedAge

Commit:
disallow mixed filter+format when using ;

Commit:
attempt to fix some affiliations bugs

tsackton commented:
Pages with bugs:
- The Wave Dancer (`Dataview: Failed to execute view '_scripts/view/get_Whereabouts.js'. TypeError: Cannot read properties of null (reading 'split')` (have not found this on other whereabouts: check override formats for errors)
- Wella Brightmoon (`Dataview: Failed to execute view '_scripts/view/get_Affiliations.js'. TypeError: Cannot read properties of undefined (reading 'matchAll')` (arises from some, but not all leader affiliations; non-leader affiliations seem fine; also fails on Kaeso, who has a plain array for an affiliation; might have to do with dates?)

Still remaining to address otherwise:
- casing only happens for the actual token, not the prefix; so e.g. Battle of Urlich Pass reports: "part of The [[Great War]]" which should probably be "Part of The [[Great War]]". The display string here is: "<(part of )partof:ty>" so maybe this is a display string bug not a code bug (e.g., should it be "<(Part of )partof:ty>"?

Commit:
update Wella Brightmoon testing page

Commit:
a few other testing header regeneration

msackton committed:
Bugfixes:

1. Add missing aNoDate to metadata
2. Allow undefined input to return "" in to "" rather than crashing
3. Move location tracking to metadatautils, so we can get rid of location
manager
4. Fix 2 whereabouts bugs:  (a) undefined format str means all formats are
ok and (b) when you have a string like "travelling to Embry" we need
to force the link text to appear because "travelling to" is treated
like link text

msackton committed:
Token parse: return "" for undefined input

Commit:
apply casing format only to prefix and suffix; see 507-516 for details

msackton commented:
> Pages with bugs:
> 
> * The Wave Dancer (`Dataview: Failed to execute view '_scripts/view/get_Whereabouts.js'. TypeError: Cannot read properties of null (reading 'split')` (have not found this on other whereabouts: check override formats for errors)
> * Wella Brightmoon (`Dataview: Failed to execute view '_scripts/view/get_Affiliations.js'. TypeError: Cannot read properties of undefined (reading 'matchAll')` (arises from some, but not all leader affiliations; non-leader affiliations seem fine; also fails on Kaeso, who has a plain array for an affiliation; might have to do with dates?)
> 

Both of these are fixed by https://github.com/tsackton/taelgar/commit/847d853bc3a35350e62cb0ab2b2a5c9b14221dde

(The issue with affiliations was the lack of an aNoDates string in metadata.json; the issue with Wave Dancer was a lack of handling an undefined format string)

tsackton commented:
> casing only happens for the actual token, not the prefix; so e.g. Battle of Urlich Pass reports: "part of The [[Great War]]" which should probably be "Part of The [[Great War]]". The display string here is: "<(part of )partof:ty>" so maybe this is a display string bug not a code bug (e.g., should it be "<(Part of )partof:ty>"?

This is fixed, at least for Battle of Urlich Pass, by [c4c0d63](https://github.com/tsackton/taelgar/pull/25/commits/c4c0d63fc41f5939119a5fd649e27d93657372fa). Currently if there are different casing rules for format and firstFormat, format always applies to the prefix. Conceptually it might be more correct to apply firstFormat to the prefix and format to the suffix? 

Alternatively, applying casing rules to prefix/suffix text might not be worth the hassle. The main impact is if you have something like "<subtypeof(,)> <(part of)partof:u>" where you don't want "part of" capitalized unless there is no subtype.

msackton committed:
Handful of page fixes

msackton committed:
More small file fixes
